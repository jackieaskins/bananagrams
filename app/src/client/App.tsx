import {
  ChakraProvider,
  Spinner,
  Text,
  ThemeConfig,
  ToastProviderProps,
  extendTheme,
} from "@chakra-ui/react";
import { useEffect, useState } from "react";
import { BrowserRouter } from "react-router-dom";
import AppRoutes from "./AppRoutes";
import LocalStorageProvider from "./LocalStorageProvider";
import BraveBrowserDialog from "./dialogs/BraveBrowserDialog";
import ServerDisconnectDialog from "./dialogs/ServerDisconnectDialog";
import CenteredLayout from "./layouts/CenteredLayout";
import NavMenu from "./menus/NavMenu";
import { NavMenuContext } from "./menus/NavMenuContext";
import { socket } from "./socket";

const config: ThemeConfig = {
  initialColorMode: "system",
  useSystemColorMode: true,
};
const theme = extendTheme({ config });

const toastOptions: ToastProviderProps = {
  defaultOptions: {
    position: "top-left",
  },
};

export default function App(): JSX.Element {
  const [renderNavMenu, setRenderNavMenu] = useState(true);
  const [hasConnected, setHasConnected] = useState(socket.connected);

  useEffect(() => {
    function handleConnect() {
      setHasConnected(true);
    }

    socket.on("connect", handleConnect);
    socket.connect();

    return () => {
      socket.off("connect", handleConnect);
    };
  }, []);

  return (
    <LocalStorageProvider>
      <ChakraProvider theme={theme} toastOptions={toastOptions}>
        {hasConnected ? (
          <NavMenuContext.Provider value={[renderNavMenu, setRenderNavMenu]}>
            <BrowserRouter>
              <ServerDisconnectDialog />
              <AppRoutes />
              {renderNavMenu && <NavMenu />}
              <BraveBrowserDialog />
            </BrowserRouter>
          </NavMenuContext.Provider>
        ) : (
          <CenteredLayout>
            <Text textAlign="center">
              <Spinner />
            </Text>
          </CenteredLayout>
        )}
      </ChakraProvider>
    </LocalStorageProvider>
  );
}
