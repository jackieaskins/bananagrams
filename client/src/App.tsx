import {
  ChakraProvider,
  ThemeConfig,
  ToastProviderProps,
  extendTheme,
} from "@chakra-ui/react";
import { BrowserRouter } from "react-router-dom";
import AppRoutes from "./AppRoutes";
import ServerDisconnectDialog from "./dialogs/ServerDisconnectDialog";
import NavMenu from "./menus/NavMenu";
import { SocketProvider } from "./socket/SocketContext";

const config: ThemeConfig = {
  initialColorMode: "system",
  useSystemColorMode: true,
};
const theme = extendTheme({ config });

const toastOptions: ToastProviderProps = {
  defaultOptions: {
    position: "bottom-left",
  },
};

export default function App(): JSX.Element {
  return (
    <ChakraProvider theme={theme} toastOptions={toastOptions}>
      <SocketProvider>
        <BrowserRouter>
          <ServerDisconnectDialog />
          <AppRoutes />
          <NavMenu />
        </BrowserRouter>
      </SocketProvider>
    </ChakraProvider>
  );
}
