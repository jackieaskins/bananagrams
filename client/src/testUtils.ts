import {
  RenderOptions,
  RenderResult,
  render as rtlRender,
} from "@testing-library/react";
import userEvent, { UserEvent } from "@testing-library/user-event";

// https://testing-library.com/docs/user-event/intro
export function render(
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">,
): RenderResult & { user: UserEvent } {
  return {
    user: userEvent.setup(),
    ...rtlRender(ui, options),
  };
}
