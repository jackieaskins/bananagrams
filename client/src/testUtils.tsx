import React from 'react';
import { shallow } from 'enzyme';

type CallbackFn = (...args: any[]) => any;

const TestHook = ({ callback }: { callback: CallbackFn }): null => {
  callback();
  return null;
};

export const testHook = (callback: CallbackFn): void => {
  shallow(<TestHook callback={callback} />);
};
