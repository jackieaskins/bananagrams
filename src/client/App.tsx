import {
  ChakraProvider,
  ThemeConfig,
  ToastProviderProps,
  extendTheme,
} from "@chakra-ui/react";
import { useState } from "react";
import { BrowserRouter } from "react-router-dom";
import AppRoutes from "./AppRoutes";
import LocalStorageProvider from "./LocalStorageProvider";
import ServerDisconnectDialog from "./dialogs/ServerDisconnectDialog";
import NavMenu from "./menus/NavMenu";
import { NavMenuContext } from "./menus/NavMenuContext";
import useSocketManager from "./socket/useSocketManager";

const config: ThemeConfig = {
  initialColorMode: "system",
  useSystemColorMode: true,
};
const theme = extendTheme({ config });

const toastOptions: ToastProviderProps = {
  defaultOptions: {
    position: "top-left",
  },
};

export default function App(): JSX.Element {
  useSocketManager();
  const [renderNavMenu, setRenderNavMenu] = useState(true);

  return (
    <LocalStorageProvider>
      <ChakraProvider theme={theme} toastOptions={toastOptions}>
        <NavMenuContext.Provider value={[renderNavMenu, setRenderNavMenu]}>
          <BrowserRouter>
            <ServerDisconnectDialog />
            <AppRoutes />
            {renderNavMenu && <NavMenu />}
          </BrowserRouter>
        </NavMenuContext.Provider>
      </ChakraProvider>
    </LocalStorageProvider>
  );
}
