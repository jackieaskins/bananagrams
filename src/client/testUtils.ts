import {
  RenderOptions,
  RenderResult,
  render as rtlRender,
} from "@testing-library/react";
import userEvent, { UserEvent } from "@testing-library/user-event";

export const CURRENT_PLAYER_ID = "currentPlayerId";

// https://testing-library.com/docs/user-event/intro
export function renderComponent(
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">,
): RenderResult & { user: UserEvent } {
  return {
    user: userEvent.setup(),
    ...rtlRender(ui, options),
  };
}
