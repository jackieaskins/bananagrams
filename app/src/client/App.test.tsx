import App from "./App";
import { renderComponent } from "./testUtils";

describe("<App />", () => {
  it("renders properly", () => {
    const { asFragment } = renderComponent(<App />);

    expect(asFragment()).toMatchSnapshot();
  });
});
