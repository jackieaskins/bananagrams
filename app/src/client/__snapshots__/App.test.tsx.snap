// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders properly 1`] = `
<DocumentFragment>
  <div
    class="css-h74a88"
  >
    <div
      class="chakra-container css-vhc7ah"
    >
      <form
        autocomplete="off"
        class="css-0"
      >
        <div
          class="chakra-stack css-ahthbn"
        >
          <h1
            class="chakra-heading css-xrxou1"
          >
            Start a new game
          </h1>
          <div
            class="chakra-stack css-8g8ihq"
          >
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-g6pte"
                for="gameName"
                id="field-:r2:-label"
              >
                Game name
                <span
                  aria-hidden="true"
                  class="chakra-form__required-indicator css-1tfjd1n"
                  role="presentation"
                >
                  *
                </span>
              </label>
              <input
                aria-required="true"
                class="chakra-input css-1cjy4zv"
                id="gameName"
                name="gameName"
                required=""
                value=""
              />
            </div>
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-g6pte"
                for="username"
                id="field-:r3:-label"
              >
                Username
                <span
                  aria-hidden="true"
                  class="chakra-form__required-indicator css-1tfjd1n"
                  role="presentation"
                >
                  *
                </span>
              </label>
              <input
                aria-required="true"
                autocomplete="username"
                class="chakra-input css-1cjy4zv"
                id="username"
                name="username"
                required=""
                value=""
              />
            </div>
          </div>
          <button
            class="chakra-button css-jut409"
            type="submit"
          >
            Create game
          </button>
        </div>
      </form>
    </div>
  </div>
  <button
    aria-controls="menu-list-:r5:"
    aria-expanded="false"
    aria-haspopup="menu"
    class="chakra-button chakra-menu__menu-button css-17ocm81"
    id="menu-button-:r5:"
    type="button"
  >
    <svg
      aria-hidden="true"
      fill="currentColor"
      focusable="false"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 448 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 96C0 78.3 14.3 64 32 64H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H32C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H32c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32H32c-17.7 0-32-14.3-32-32s14.3-32 32-32H416c17.7 0 32 14.3 32 32z"
      />
    </svg>
  </button>
  <div
    class="css-r6z5ec"
    style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
  >
    <div
      aria-orientation="vertical"
      class="chakra-menu__menu-list css-1kfu8nn"
      id="menu-list-:r5:"
      role="menu"
      style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
      tabindex="-1"
    >
      <a
        class="chakra-menu__menuitem css-18esm8n"
        data-index="0"
        href="/changelog"
        id="menu-list-:r5:-menuitem-:r6:"
        role="menuitem"
        tabindex="-1"
      >
        <span
          class="chakra-menu__icon-wrapper css-7halv9"
        >
          <svg
            aria-hidden="true"
            class="chakra-menu__icon"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M40 48C26.7 48 16 58.7 16 72v48c0 13.3 10.7 24 24 24H88c13.3 0 24-10.7 24-24V72c0-13.3-10.7-24-24-24H40zM192 64c-17.7 0-32 14.3-32 32s14.3 32 32 32H480c17.7 0 32-14.3 32-32s-14.3-32-32-32H192zm0 160c-17.7 0-32 14.3-32 32s14.3 32 32 32H480c17.7 0 32-14.3 32-32s-14.3-32-32-32H192zm0 160c-17.7 0-32 14.3-32 32s14.3 32 32 32H480c17.7 0 32-14.3 32-32s-14.3-32-32-32H192zM16 232v48c0 13.3 10.7 24 24 24H88c13.3 0 24-10.7 24-24V232c0-13.3-10.7-24-24-24H40c-13.3 0-24 10.7-24 24zM40 368c-13.3 0-24 10.7-24 24v48c0 13.3 10.7 24 24 24H88c13.3 0 24-10.7 24-24V392c0-13.3-10.7-24-24-24H40z"
            />
          </svg>
        </span>
        <span
          style="pointer-events: none; flex: 1;"
        >
          View changelog
        </span>
      </a>
      <hr
        aria-orientation="horizontal"
        class="chakra-menu__divider css-12vdn2q"
      />
      <a
        class="chakra-menu__menuitem css-18esm8n"
        data-index="1"
        href="/tutorial"
        id="menu-list-:r5:-menuitem-:r7:"
        role="menuitem"
        tabindex="-1"
      >
        <span
          class="chakra-menu__icon-wrapper css-7halv9"
        >
          <svg
            aria-hidden="true"
            class="chakra-menu__icon"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 640 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M320 32c-8.1 0-16.1 1.4-23.7 4.1L15.8 137.4C6.3 140.9 0 149.9 0 160s6.3 19.1 15.8 22.6l57.9 20.9C57.3 229.3 48 259.8 48 291.9v28.1c0 28.4-10.8 57.7-22.3 80.8c-6.5 13-13.9 25.8-22.5 37.6C0 442.7-.9 448.3 .9 453.4s6 8.9 11.2 10.2l64 16c4.2 1.1 8.7 .3 12.4-2s6.3-6.1 7.1-10.4c8.6-42.8 4.3-81.2-2.1-108.7C90.3 344.3 86 329.8 80 316.5V291.9c0-30.2 10.2-58.7 27.9-81.5c12.9-15.5 29.6-28 49.2-35.7l157-61.7c8.2-3.2 17.5 .8 20.7 9s-.8 17.5-9 20.7l-157 61.7c-12.4 4.9-23.3 12.4-32.2 21.6l159.6 57.6c7.6 2.7 15.6 4.1 23.7 4.1s16.1-1.4 23.7-4.1L624.2 182.6c9.5-3.4 15.8-12.5 15.8-22.6s-6.3-19.1-15.8-22.6L343.7 36.1C336.1 33.4 328.1 32 320 32zM128 408c0 35.3 86 72 192 72s192-36.7 192-72L496.7 262.6 354.5 314c-11.1 4-22.8 6-34.5 6s-23.5-2-34.5-6L143.3 262.6 128 408z"
            />
          </svg>
        </span>
        <span
          style="pointer-events: none; flex: 1;"
        >
          Play tutorial
        </span>
      </a>
      <a
        class="chakra-menu__menuitem css-18esm8n"
        data-index="2"
        href="https://bananagrams.com/blogs/news/how-to-play-bananagrams-instructions-for-getting-started"
        id="menu-list-:r5:-menuitem-:r8:"
        rel="noreferrer"
        role="menuitem"
        tabindex="-1"
        target="_blank"
      >
        <span
          class="chakra-menu__icon-wrapper css-7halv9"
        >
          <svg
            aria-hidden="true"
            class="chakra-menu__icon"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M249.6 471.5c10.8 3.8 22.4-4.1 22.4-15.5V78.6c0-4.2-1.6-8.4-5-11C247.4 52 202.4 32 144 32C93.5 32 46.3 45.3 18.1 56.1C6.8 60.5 0 71.7 0 83.8V454.1c0 11.9 12.8 20.2 24.1 16.5C55.6 460.1 105.5 448 144 448c33.9 0 79 14 105.6 23.5zm76.8 0C353 462 398.1 448 432 448c38.5 0 88.4 12.1 119.9 22.6c11.3 3.8 24.1-4.6 24.1-16.5V83.8c0-12.1-6.8-23.3-18.1-27.6C529.7 45.3 482.5 32 432 32c-58.4 0-103.4 20-123 35.6c-3.3 2.6-5 6.8-5 11V456c0 11.4 11.7 19.3 22.4 15.5z"
            />
          </svg>
        </span>
        <span
          style="pointer-events: none; flex: 1;"
        >
          Read official instructions
        </span>
      </a>
      <hr
        aria-orientation="horizontal"
        class="chakra-menu__divider css-12vdn2q"
      />
      <a
        class="chakra-menu__menuitem css-18esm8n"
        data-index="3"
        href="https://github.com/jackieaskins/bananagrams/issues/new?template=bug_report.md"
        id="menu-list-:r5:-menuitem-:r9:"
        rel="noreferrer"
        role="menuitem"
        tabindex="-1"
        target="_blank"
      >
        <span
          class="chakra-menu__icon-wrapper css-7halv9"
        >
          <svg
            aria-hidden="true"
            class="chakra-menu__icon"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 0c53 0 96 43 96 96v3.6c0 15.7-12.7 28.4-28.4 28.4H188.4c-15.7 0-28.4-12.7-28.4-28.4V96c0-53 43-96 96-96zM41.4 105.4c12.5-12.5 32.8-12.5 45.3 0l64 64c.7 .7 1.3 1.4 1.9 2.1c14.2-7.3 30.4-11.4 47.5-11.4H312c17.1 0 33.2 4.1 47.5 11.4c.6-.7 1.2-1.4 1.9-2.1l64-64c12.5-12.5 32.8-12.5 45.3 0s12.5 32.8 0 45.3l-64 64c-.7 .7-1.4 1.3-2.1 1.9c6.2 12 10.1 25.3 11.1 39.5H480c17.7 0 32 14.3 32 32s-14.3 32-32 32H416c0 24.6-5.5 47.8-15.4 68.6c2.2 1.3 4.2 2.9 6 4.8l64 64c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0l-63.1-63.1c-24.5 21.8-55.8 36.2-90.3 39.6V240c0-8.8-7.2-16-16-16s-16 7.2-16 16V479.2c-34.5-3.4-65.8-17.8-90.3-39.6L86.6 502.6c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3l64-64c1.9-1.9 3.9-3.4 6-4.8C101.5 367.8 96 344.6 96 320H32c-17.7 0-32-14.3-32-32s14.3-32 32-32H96.3c1.1-14.1 5-27.5 11.1-39.5c-.7-.6-1.4-1.2-2.1-1.9l-64-64c-12.5-12.5-12.5-32.8 0-45.3z"
            />
          </svg>
        </span>
        <span
          style="pointer-events: none; flex: 1;"
        >
          Report bug
        </span>
      </a>
      <a
        class="chakra-menu__menuitem css-18esm8n"
        data-index="4"
        href="https://github.com/jackieaskins/bananagrams/issues/new?template=feature_request.md"
        id="menu-list-:r5:-menuitem-:ra:"
        rel="noreferrer"
        role="menuitem"
        tabindex="-1"
        target="_blank"
      >
        <span
          class="chakra-menu__icon-wrapper css-7halv9"
        >
          <svg
            aria-hidden="true"
            class="chakra-menu__icon"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 384 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M272 384c9.6-31.9 29.5-59.1 49.2-86.2l0 0c5.2-7.1 10.4-14.2 15.4-21.4c19.8-28.5 31.4-63 31.4-100.3C368 78.8 289.2 0 192 0S16 78.8 16 176c0 37.3 11.6 71.9 31.4 100.3c5 7.2 10.2 14.3 15.4 21.4l0 0c19.8 27.1 39.7 54.4 49.2 86.2H272zM192 512c44.2 0 80-35.8 80-80V416H112v16c0 44.2 35.8 80 80 80zM112 176c0 8.8-7.2 16-16 16s-16-7.2-16-16c0-61.9 50.1-112 112-112c8.8 0 16 7.2 16 16s-7.2 16-16 16c-44.2 0-80 35.8-80 80z"
            />
          </svg>
        </span>
        <span
          style="pointer-events: none; flex: 1;"
        >
          Suggest new feature
        </span>
      </a>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
